shell/platform/android/io/flutter/embedding/engine/systemchannels/RestorationChannel.java

```diff
@@ -0,0 +1,153 @@
+ *
+ * <p>The channel can be configured to delay responding to the framework's request for restoration
+ * data via {@code waitForRestorationData} until the engine-side has provided the data. This is
+ * useful when the engine is pre-warmed at a point in the application's life cycle where the
+ * restoration data is not available yet. For example, if the engine is pre-warmed as part of the
+ * Application before an Activity is created, this flag should be set to true because Android will
+ * only provide the restoration data to the Activity during the onCreate callback.
+ *
+ * <p>The current restoration data provided by the framework can be read via {@code
+ * getRestorationData()}.
+ */
+public class RestorationChannel {
+  private static final String TAG = "RestorationChannel";
+
+  public RestorationChannel(
+      @NonNull DartExecutor dartExecutor, @NonNull boolean waitForRestorationData) {
+    this(
+        new MethodChannel(dartExecutor, "flutter/restoration", StandardMethodCodec.INSTANCE),
+        waitForRestorationData);
+  }
+
+  RestorationChannel(MethodChannel channel, @NonNull boolean waitForRestorationData) {
+    this.channel = channel;
+    this.waitForRestorationData = waitForRestorationData;
+
+    channel.setMethodCallHandler(handler);
+  }
+
+  /**
+   * Whether the channel delays responding to the framework's initial request for restoration data
+   * until {@code setRestorationData} has been called.
+   *
+   * <p>If the engine never calls {@code setRestorationData} this flag must be set to false. If set
+   * to true, the engine must call {@code setRestorationData} either with the actual restoration
+   * data as argument or null if it turns out that there is no restoration data.
+   *
+   * <p>If the response to the framework's request for restoration data is not delayed until the
+   * data has been set via {@code setRestorationData}, the framework may intermittently initialize
+   * itself to default values until the restoration data has been made available. Setting this flag
+   * to true avoids that extra work.
+   */
+  public final boolean waitForRestorationData;
+
+  // Holds the the most current restoration data which may have been provided by the engine
+  // via "setRestorationData" or by the framework via the method channel. This is the data the
+  // framework should be restored to in case the app is terminated.
+  private byte[] restorationData;
+  private MethodChannel channel;
+  private MethodChannel.Result pendingFrameworkRequest;
+  private boolean engineHasProvidedData = false;
+  private boolean frameworkHasRequestedData = false;
+
+  /** Obtain the most current restoration data that the framework has provided. */
+  public byte[] getRestorationData() {
+    return restorationData;
+  }
+
+  /** Set the restoration data from which the framework will restore its state. */
+  public void setRestorationData(byte[] data) {
+    engineHasProvidedData = true;
+    if (pendingFrameworkRequest != null) {
```

Comment:
ah I misread this to be a bool. Maybe "pendingFrameworkRestorationChannelRequest"?